apps/alb_monitor/lib/alb_monitor/monitor.ex

```diff
@@ -0,0 +1,94 @@
+defmodule ALBMonitor.Monitor do
+  @moduledoc "See `ALBMonitor`."
+
+  use GenServer
+
+  alias ExAws.ElasticLoadBalancingV2, as: ALB
+
+  require Logger
+
+  @ex_aws Application.compile_env!(:alb_monitor, :ex_aws)
+  @http Application.compile_env!(:alb_monitor, :http)
+
+  defmodule State do
+    @moduledoc "State of the GenServer."
+    defstruct check_interval: 5_000,
+              fetch_env_fn: &System.fetch_env/1,
+              instance_ip: nil,
+              shutdown_fn: &:init.stop/0
+  end
+
+  @spec start_link(%State{}) :: GenServer.on_start()
+  def start_link(initial_state \\ %State{}) do
+    GenServer.start_link(__MODULE__, initial_state)
+  end
+
+  @impl true
+  def init(state) do
+    schedule_check(state)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_info(:check, %State{instance_ip: nil} = state) do
+    new_state = %{state | instance_ip: get_instance_ip(state)}
+    schedule_check(new_state)
+    {:noreply, new_state}
+  end
+
+  @impl true
+  def handle_info(:check, %State{shutdown_fn: shutdown} = state) do
+    case get_instance_health(state) do
+      "draining" ->
+        log_info("shutdown")
+        shutdown.()
+        {:stop, nil, state}
+
+      _ ->
+        schedule_check(state)
+        {:noreply, state}
+    end
+  end
+
+  # https://github.com/benoitc/hackney/issues/464
+  @impl true
+  def handle_info({:ssl_closed, _}, state), do: {:noreply, state}
+
+  defp get_instance_health(%State{fetch_env_fn: fetch_env, instance_ip: instance_ip}) do
+    with {:ok, target_group} <- fetch_env.("ALB_TARGET_GROUP_ARN"),
+         {:ok, %{body: %{target_health_descriptions: health_descs}}} <-
+           target_group |> ALB.describe_target_health() |> @ex_aws.request(),
+         %{target_health: health} <-
+           Enum.find(health_descs, &match?(%{targets: [%{id: ^instance_ip}]}, &1)) do
+      health
+    else
+      unmatched ->
+        log_warn("get_instance_health failed: #{inspect(unmatched)}")
+        nil
+    end
+  end
+
+  defp get_instance_ip(%State{fetch_env_fn: fetch_env}) do
+    with {:ok, uri} <- fetch_env.("ECS_CONTAINER_METADATA_URI"),
+         {:ok, %{body: body}} <- @http.get(uri),
+         {:ok, %{"Networks" => [%{"IPv4Addresses" => [ip_address]}]}} <- Poison.decode(body) do
+      ip_address
+    else
+      :error ->
+        # the `fetch_env` failed, likely not running on ECS
+        log_info("no_ecs_metadata_uri")
+        nil
+
+      other ->
+        log_warn("get_instance_ip failed: #{inspect(other)}")
+        nil
+    end
+  end
+
+  defp log_warn(message) when is_binary(message), do: Logger.warn("alb_monitor " <> message)
+  defp log_info(message) when is_binary(message), do: Logger.info("alb_monitor " <> message)
+
+  defp schedule_check(%State{check_interval: check_interval}) do
+    Process.send_after(self(), :check, check_interval)
```

Comment:
Rather than manually doing this, it might be easier to use [:timer.send_interval](https://erlang.org/doc/man/timer.html#send_interval-2).

Reply:
I'm a bit wary of `send_interval` since it doesn't consider the time taken to handle the messages, creating the possibility of filling up the mailbox or sending excessive requests to an external service (imagine if the ALB API starts timing out or responding very slowly). With this approach it's not possible for a `check` to be less than 5 seconds after the previous one, no matter how long it took to perform the check, so it seems safer. Does that make sense?